devmand/gateway/src/devmand/channels/cli/SshSessionAsync.cpp

```diff
@@ -66,9 +66,12 @@ Future<Unit> SshSessionAsync::write(const string& command) {
 }
 
 Future<string> SshSessionAsync::readUntilOutput(const string& lastOutput) {
-  return via(executor.get(), [this, lastOutput] {
-    return this->readUntilOutputBlocking(lastOutput);
-  });
+  this->readingState.currentLastOutput = lastOutput;
+  this->readingState.promise = std::make_shared<Promise<string>>();
+  this->readingState.outputSoFar = "";
+  processDataInBuffer(); // we could have had something already waiting in the
```

Comment:
we could have had something already waiting ... is that true ? Because if something is read, it is inserted to queue and then immediatelly poped in the processDataInBuffer. So queue is empty, we here remove all possible state in output so far, so we are losing any data coming before readUntilOutput is called.

One more thing, the processDataInBuffer is triggerred from multiple places and invoked concurently in the executor. Is the access to readingState protected from concurrent access ?

Reply:
I thought so to, but the queue might already be populated. The sequence is as follows:

1. write command to SSH (PromptAwareCli call)
2. NIO already writes response to queue     <------ here we already get the data
3. readuntiloutput triggered  (from PromptAwareCli call)

"Is the access to readingState protected from concurrent access ?"
You are right and it is wrong, I found out a couple of hours ago, it is not protected and it won't be, I will need to use a serialExecutor. I will do an update shortly, I am checking some corner cases